infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/UniqueWorkspacePVCStrategy.java

```diff
@@ -156,42 +180,100 @@ public void cleanup(Workspace workspace) throws InfrastructureException {
         .delete(ImmutableMap.of(CHE_WORKSPACE_ID_LABEL, workspaceId));
   }
 
+  private void fillInExistingPVCs(KubernetesEnvironment k8sEnv, String workspaceId)
+      throws InfrastructureException {
+    k8sEnv
+        .getPersistentVolumeClaims()
+        .putAll(
+            factory
+                .create(workspaceId)
+                .persistentVolumeClaims()
+                .getByLabel(CHE_WORKSPACE_ID_LABEL, workspaceId)
+                .stream()
+                .peek(pvc -> pvc.getAdditionalProperties().put(CHE_PROVISIONED_PVC_PROPERTY, true))
+                .collect(toMap(pvc -> pvc.getMetadata().getName(), Function.identity())));
+  }
+
   private void provisionCheVolumes(
-      Map<String, PersistentVolumeClaim> volumeName2PVC,
       KubernetesEnvironment k8sEnv,
-      String workspaceId)
-      throws InfrastructureException {
+      String workspaceId,
+      Map<String, PersistentVolumeClaim> volumeName2PVC) {
     for (PodData pod : k8sEnv.getPodsData().values()) {
       final PodSpec podSpec = pod.getSpec();
       List<Container> containers = new ArrayList<>();
       containers.addAll(podSpec.getContainers());
       containers.addAll(podSpec.getInitContainers());
       for (Container container : containers) {
         final String machineName = Names.machineName(pod, container);
-        Map<String, Volume> volumes = k8sEnv.getMachines().get(machineName).getVolumes();
-        addMachineVolumes(
-            workspaceId,
-            k8sEnv.getPersistentVolumeClaims(),
-            volumeName2PVC,
-            pod,
-            container,
-            volumes);
+        InternalMachineConfig machineConfig = k8sEnv.getMachines().get(machineName);
+        if (machineConfig == null) {
+          continue;
+        }
+        Map<String, Volume> volumes = machineConfig.getVolumes();
+        addMachineVolumes(workspaceId, k8sEnv, volumeName2PVC, pod, container, volumes);
+      }
+    }
+  }
+
+  private void processUserDefinedPVCs(
```

Comment:
Could you add javadoc for this method? It's quite complicated in terms of how it changes the state, and `process` doesn't really explain what is being done.

Reply:
User-defined PVC processing is already described in class java docs. Added short javadoc for this private method that references class javadoc. See https://github.com/eclipse/che/pull/12335/commits/7b492cf47009b70965bcd7a7a4ad7ec3803e6f9d
Do you think it is OK enough?

Reply:
That makes sense, thanks.